### PR TITLE
[READY][MOSTLY MODULAR] Mediguns!

### PR DESCRIPTION
**Pull request by softcerv**
## About The Pull Request
Adds a new medical tool, the medigun. This is meant to serve as a more balanced and versatile medical tool order-able from cargo in-place of the medibeam. _**The original medibeam will not be removed from the game.**_ This medigun is heavily based off the one from Virgo. 

Huge thanks to @Arctaisia, who helped me make most of the textures for this PR. 

This PR is the framework for how I want mediguns to function, things are likely to change in the future as I figure out what stuff needs to be changed and added. Please leave any suggestions or changes in the comments here or DM me on discord at Bunny#5839.
## How it works
Mediguns fire a shot of healing energy the heals the target based on the selected damage type. The healing types are added to the gun by loading medicells into their chamber. When a cell is removed, the gun will lose that firing mode, guns are only able to carry a limited amount of cells at once. Each medicell corresponds with a different type and strength of healing. All of the guns come with a basic oxygen heal mode built in. Aside from the medicell/healing functionality and a longer charge time, they function similarly to other energy guns. The mediguns charge significantly slower than other energy weapons to encourage the user to think about their shots more, instead of being able to endlessly pump healing shots into someone. 

**TLDR**: This works like a energy gun, but instead of hurting, it heals and the different firing modes are added though inserting cells into it.
 
## Mediguns
So far, I have three different variants of the gun. 
1. CWM-479. This is the standard model that can be ordered from cargo, it has slots for three medicells and has a horrible recharge rate. This can be acquired through cargo for around 3000cr, Tier 1 medicells are included.
- Heals 100 damage with a full charge using Tier I Cells

image 
2. CWM-479-FC. An upgraded version of the standard medigun with a increased charge rate and higher cell capacity. This will be obtained from an upgrade kit locked behind research. 
- Heals 125 damage with a full charge using Tier I Cells

image
3. CWM-479-CC. The CMO's medigun, it can fit six medicells and features a large cell capacity. The gun self charges, but it shares the same recharge rate as the standard medigun and cannot be placed in chargers. 
- Heals 150 damage with a full charge using Tier I Cells

image

## Medicells
There will be three different tiers. I might look into adding mixture cells in the future too. 

- Tier I ((comes with the mediguns))
image
  - Brute and Burn heal 5 damage per shot
  - Toxin heals 2.5 damage and 20 rads per shot
- Tier II is locked behind advanced bio research. Shots from this tier and the one above will be able to pass through glass.
image
  - Brute and Burn heal 7.5 damage per shot
  - Toxin heals 3.5 damage and 30 rads per shot
  - Oxygen heals 10 damage per shot
- Tier III is locked behind alien medical research
image
  - Brute and Burn heal 10 damage per shot
  - Toxin heals 5 and 40 rads damage per shot
  - Oxygen heals 15 damage per shot.

## Things not included here but that I want to do with future PRs relating to mediguns
- A special variant of the Medigun for the sec medic more focused around combat healing and providing support.
- Implantable Medigun.
- More cell types that could possibly be unlocked through BEPIS, like mixture healing or more specialized effects.
- Possibly more upgrade kits
- Traitor Medigun and Medigun Cells.
- Maint/Craftable Medigun with special cells.
- Custom cases.

## TODO

- [x] Create the upgrade kit
- [x] Make a case with the gun
- [x] Make the gun purchasable from cargo
- [x] Make the upgrade kit and higher tier cells accessible through research.
- [x] Add the CMO variant to the CMO's locker
- [x] Add in textures for the guns
- [x] Add in working ammo selection textures
- [x] Add in textures for the other tiers of cells
- [x] Tweak the cells and guns a bit

## Why It's Good For The Game

This gives medical another tool to use that rewards medical staff for doing bounties, while being more balanced and versatile then the medibeam was. 
## Changelog
:cl:
imageadd: Adds in textures for mediguns and medicells.
add: Adds mediguns, upgradeable healing guns, orderable via cargo.
add:  Adds medicells, upgrades to mediguns that unlock new healing types.  
/:cl:

### CODE REVIEW

**Comment by DuffCreeper** [comment]
poggers?

File: modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	name = "Oxygen III Medicell"
+	desc = "A small cell with a intense blue glow. Can be used on Mediguns to unlock the Oxygen II Functoinality"
+	w_class = WEIGHT_CLASS_SMALL
+	icon_state = "Oxygen1"
+	ammo_type = /obj/item/ammo_casing/energy/medical/oxy3
+//End of Tier III
+//Medigun Upgrade//
+/obj/item/gun/energy/medigun/attackby(obj/item/medicell/M, mob/user)
+	if(cellcount >= maxcells)
+		to_chat(usr, span_notice("The Medigun is full, take a cell out to make room"))
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm]:
```suggestion
		to_chat(user, span_notice("The Medigun is full, take a cell out to make room"))
```

File: modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	else
+		if(!user.transferItemToLoc(M, src))
+			return
+		playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm]:
```suggestion
		playsound(loc, 'sound/machines/click.ogg', 50, 1)
```

File: modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		to_chat(user, span_notice("You install the medicell."))
+		ammo_type += new M.ammo_type(src)
+		installedcells += M
+		cellcount += 1
+
+/obj/item/gun/energy/medigun/RightClick(mob/user)
+	if(cellcount >= 1)
+		to_chat(usr, span_notice("You remove a cell"))
+		installedcells[installedcells.len].forceMove(drop_location())
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm]:
```suggestion
		var/obj/item/last_cell = installedcells[installedcells.len]
		if(last_cell)
			last_cell.forceMove(drop_location())
		installedcells -= last_cell
```

File: modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+		installedcells.len--
+		ammo_type.len--
+		cellcount -= 1
+	else
+		to_chat(usr, span_notice("The Medigun has no cells inside"))
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm]:
```suggestion
		to_chat(user, span_notice("The Medigun has no cells inside"))
```

File: modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+			return
+		playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
+		to_chat(user, span_notice("You install the medicell."))
+		ammo_type += new M.ammo_type(src)
+		installedcells += M
+		cellcount += 1
+
+/obj/item/gun/energy/medigun/RightClick(mob/user)
+	if(cellcount >= 1)
+		to_chat(usr, span_notice("You remove a cell"))
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm]:
```suggestion
		to_chat(user, span_notice("You remove a cell"))
```

File: modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+	var/list/installedcells = list()
+
+/obj/item/gun/energy/medigun/examine(mob/user)
+	. = ..()
+	if(maxcells)
+		. += "<b>[cellcount]</b> out of <b>[maxcells]</b> cell slots are filled."
+		. += span_info("You can use right click to remove the most recently inserted Medicell from the chamber.")
+		for(var/A in installedcells)
+			var/obj/item/medicell/M = A
+			. += span_notice("There is \a [M] in the chamber.")
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm]:
```suggestion
		for(var/cell in installedcells)
			var/obj/item/medicell/medicell = cell
			. += span_notice("There is \a [medicell] in the chamber.")
```

File: modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medcells.dm [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+	name = "Basic Medigun Cell"
+	maxcharge = 1200
+	chargerate = 80
+
+obj/item/stock_parts/cell/medigun/upgraded
+	name = "Upgraded Medigun Cell"
+	maxcharge = 1500
+	chargerate = 160
+
+obj/item/stock_parts/cell/medigun/experimental //This cell type is meant to be used in self charging mediguns like CMO and ERT one.//
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medcells.dm]:
```suggestion
/obj/item/stock_parts/cell/medigun/experimental //This cell type is meant to be used in self charging mediguns like CMO and ERT one.//
```

File: modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medcells.dm [hunk 8]
```diff
@@ -0,0 +1,163 @@
+//Medigun Cells/
+obj/item/stock_parts/cell/medigun/ //This is the cell that mediguns from cargo will come with//
+	name = "Basic Medigun Cell"
+	maxcharge = 1200
+	chargerate = 80
+
+obj/item/stock_parts/cell/medigun/upgraded
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medcells.dm]:
```suggestion
/obj/item/stock_parts/cell/medigun/upgraded
```

File: modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medcells.dm [hunk 9]
```diff
@@ -0,0 +1,163 @@
+//Medigun Cells/
+obj/item/stock_parts/cell/medigun/ //This is the cell that mediguns from cargo will come with//
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medcells.dm]:
```suggestion
/obj/item/stock_parts/cell/medigun/ //This is the cell that mediguns from cargo will come with//
```

**Review by ZephyrTFA — CHANGES_REQUESTED** [review]

File: modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+		cellcount += 1
+
+/obj/item/gun/energy/medigun/RightClick(mob/user)
+	if(cellcount >= 1)
+		to_chat(user, span_notice("You remove a cell"))
+		var/obj/item/last_cell = installedcells[installedcells.len]
+		if(last_cell)
+			last_cell.forceMove(drop_location())
+		installedcells -= last_cell
+		installedcells.len--
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm]:
```suggestion
```
Didn't see this line; thats my fault

Comment (softcerv) [reply modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm]: ah, it's good. Thank you for reviewing my code by the way!

**Comment by FlamingLily** [comment]
image
Let's goooooooooooooooooo

File: modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+	ammo_type = list(/obj/item/ammo_casing/energy/medical) //The default option that heals Oxygen//
+	ammo_x_offset = 2
+	w_class = WEIGHT_CLASS_NORMAL
+	cell_type = /obj/item/stock_parts/cell/medigun/
+	ammo_x_offset = 2
+	charge_sections = 3
+	has_gun_safety = TRUE
+	var/maxcells = 3
+	var/cellcount = 0
+	var/list/installedcells = list()
```
Comment (MLGTASTICa) [inline modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm:17]: You can get rid of installedcells and use the guns contents list , instead of making a new one , and for the cell count you could could just check its length

Comment (MLGTASTICa) [reply modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm:17]: no pressure to do it however , it saves very little in terms of performance

**Comment by lyricalpaws** [comment]
Giving medical guns that heal? That's very good, yes please. (Also as someone who used to play virgo med and have tried this out, they're actually pretty nicely balanced.)

File: modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm (hunk 11 above)
Comment (softcerv) [reply modular_skyrat/modules/modular_weapons/code/modules/projectiles/guns/energy/medgun/medguns.dm:17]: Aye, I might look into that in a bit. Thank you for the suggestion!

**Comment by softcerv** [comment]
> 
> 
> Giving medical guns that heal? That's very good, yes please. (Also as someone who used to play virgo med and have tried this out, they're actually pretty nicely balanced.)

I could totally see adding the mixture and special cells from Virgo in a future PR depending on how this one goes! Hopefully it should feel kinda like the one on Virgo but more suited towards TG code!

**Comment by ZephyrTFA** [comment]
> 
> 
> This PR has been inactive for long enough to be automatically marked as stale. This means it is at risk of being auto closed in ~ 3 days, please address any outstanding review items and ensure your PR is finished, if these are all true and you are auto-staled anyway, you need to actively ask maintainers if your PR will be merged. Once you have done any of the previous actions then you should request a maintainer remove the stale label on your PR, to reset the stale timer. If you feel no maintainer will respond in that time, you may wish to close this PR youself, while you seek maintainer comment, as you will then be able to reopen the PR yourself

:shoo:

**Comment by softcerv** [comment]
> 
> 
> This PR has been inactive for long enough to be automatically marked as stale. This means it is at risk of being auto closed in ~ 3 days, please address any outstanding review items and ensure your PR is finished, if these are all true and you are auto-staled anyway, you need to actively ask maintainers if your PR will be merged. Once you have done any of the previous actions then you should request a maintainer remove the stale label on your PR, to reset the stale timer. If you feel no maintainer will respond in that time, you may wish to close this PR youself, while you seek maintainer comment, as you will then be able to reopen the PR yourself

Heck, this is still being worked on!

**Comment by Funce** [comment]
Begone Stale Button

**Comment by softcerv** [comment]
Going to close this until textures get sorted out so that this isn't marked as stale.

**Comment by softcerv** [comment]
Textures are done, time to finish this PR!

**Comment by RatFromTheJungle** [comment]
image

**Comment by softcerv** [comment]
I'd recommend testmerging this first just to make sure that there aren't any major balance or other issues before it gets full on merged.

**Review by Gandalf2k15 — APPROVED** [review]
